File: examples/inbox/native/ios/InboxExtension.swift
import UIKit

public enum InboxExtension: PNPlugin {
    public static func register(into registry: PNRegistry) {
        registry.registerComponent("InboxBadge") { InboxBadgeManager() }
        registry.registerModule(InboxToolsModule<InboxToolsService>.self)
    }
}

private final class InboxBadgeManager: PNComponentManager {
    override func makeView(props: [String: Any]) -> UIView {
        let label = UILabel()
        label.font = UIFont.preferredFont(forTextStyle: .caption1)
        label.adjustsFontForContentSizeCategory = true
        label.textColor = .secondaryLabel
        return label
    }
    override func apply(view: UIView, props: [String: Any], initial: Bool) {
        guard let label = view as? UILabel, let values = try? InboxBadgeProps(PNViewState.existing(for: view)?.props ?? props) else { return }
        label.text = "\(values.count ?? 0) offline records"
    }
}

private final class InboxToolsService: InboxToolsImplementation {
    init() {}
    func format_count(count: Int) throws -> String { "\(count) issues" }
    func ready(completion: @escaping (Result<String, Error>) -> Void) {
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.01) { completion(.success("Offline ready")) }
    }
}

File: examples/inbox/native/android/dev/pythonnative/inbox/extension/InboxExtension.kt
package dev.pythonnative.inbox.extension

import android.content.Context
import android.view.View
import android.widget.TextView
import android.os.Handler
import android.os.Looper
import com.pythonnative.generated.InboxBadgeProps
import com.pythonnative.generated.InboxToolsImplementation
import com.pythonnative.generated.InboxToolsModule
import com.pythonnative.runtime.bridge.PNPlugin
import com.pythonnative.runtime.bridge.PNRegistry
import com.pythonnative.runtime.components.ComponentManager
import org.json.JSONObject

object InboxExtension: PNPlugin {
    override fun register(registry: PNRegistry) {
        registry.registerComponent("InboxBadge") { InboxBadgeManager() }
        registry.registerModule { InboxToolsModule(InboxToolsService()) }
    }
}

private class InboxBadgeManager: ComponentManager() {
    override fun createView(context: Context, tag: Long, props: JSONObject): View = TextView(context).apply { textSize = 12f }
    override fun applyProps(view: View, props: JSONObject, initial: Boolean) {
        val values = InboxBadgeProps(propsOf(view))
        (view as TextView).text = "${values.count ?: 0} offline records"
    }
}

private class InboxToolsService: InboxToolsImplementation {
    override fun format_count(count: Int): String = "$count issues"
    override fun ready(completion: (Result<String>) -> Unit) {
        Handler(Looper.getMainLooper()).postDelayed({ completion(Result.success("Offline ready")) }, 10)
    }
}
